### ISSUE TITLE
Add a _malloca-like API Span<T>.Alloc(int length)

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/runtime#25423]
(This is distinct from the proposal at https://github.com/dotnet/corefx/issues/26954, which considers a more powerful API that can give the caller specific information about the current stack. The proposal under consideration here is only for a "[KISS](https://en.wikipedia.org/wiki/KISS_principle)" API. Both proposals could exist side-by-side depending on the scenario, and in fact this proposal could depend on the more powerful API if we decide that's the right behavior.)

As the pattern `Span<T> span = stackalloc T[len];` becomes more commonplace, we're seeing a rise in incorrect usage where developers are piping through untrusted _len_ values, potentially making their applications susceptible to reliability issues or DoS attacks. The __safe__ language construct is `Span<T> span = ((uint)len <= 128 /* or other const */) ? stackalloc T[len] : new T[len];`. We can't rely on our larger developer audience remembering the safe pattern, and even our more knowledgeable advanced developers can forget the check from time to time.

Following are concrete proposals from the security team as to how this can be addressed.

First, create a utility method `public static Span<T> Span<T>.Alloc(int length)`. This API will `stackalloc` in the current frame if appropriate; otherwise the method will heap-alloc, similar to the CRT's [_malloca](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/malloca) API. Adding this method and getting the desired runtime behavior would not require compiler changes, but it would require JIT changes. The implementation of the method would default simply to `new T[length]` if _length_ is negative or is too large. Advanced developers who want to pull from an existing array pool as fallback wouldn't use this API; they'd perform the check manually.

Second, once the utility method is in place, the compiler should be smart enough to treat `return Span<T>.Alloc(...);` as a compile-time error, just as it does today for `Span<T> span = stackalloc T[...]; return span;`. This _would_ require a compiler change.

Third, the developer should be warned when passing a non-constant and non-bounds-checked value into `stackalloc`, and the warning should direct the developer to use `Span<T>.Alloc` instead. The developer should also be warned when using `stackalloc` or `Span<T>.Alloc` in a loop. (We considered warning on recursive method calls, but that really belongs in a different analyzer.) Making this a proper compiler warning is probably a bit too noisy, so it could be a Roslyn code analyzer instead. This wouldn't require a compiler or JIT change; it'd just be a new version of the analyzers package for developers who run it as part of build.

Finally, remember that this is a KISS proposal. It's intended for the majority use case for developers who may want to consume `Span`-based APIs but need a bit of a safety net. Advanced developers could still use the other APIs under consideration or perform their own custom checks as appropriate for their applications.

**Comment by tannergooding** [comment]
Just as a note, the threshold for [_malloca](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/malloca?view=vs-2019) on Windows is `#define _ALLOCA_S_THRESHOLD 1024`. While the same size may be too big/small for managed code or for non-Windows platforms, it at least gives a rough starting point.

For reference: `double.Parse` currently does a 769 byte stack allocation for its purposes and so around this may be reasonable for non-recursive methods.

**Comment by GrabYourPitchforks** [comment]
I experimented with this a little bit recently, and from the runtime perspective it turns out we can make it work without too much pain. The biggest problem is what Jared pointed out earlier: the C# compiler won't restrict these spans from escaping out of the local method. So if you use it you have to be disciplined about it without relying on compiler support. At the moment that seems strictly worse.

**Comment by tannergooding** [comment]
> At the moment that seems strictly worse.

Agreed.

**Comment by jkotas** [comment]
https://github.com/dotnet/csharplang/issues/1130 ?

**Comment by benaadams** [comment]
Also stackalloc ref types to `Span<T>` receivers https://github.com/dotnet/runtime/issues/46104#issuecomment-745633974 and https://github.com/dotnet/runtime/issues/38677 (the array rather than the objects themselves); rather than having to use ArrayPool for small object Span/Arrays that are then passed to another method (issue with for `params Span<T>` also?)

**Comment by GrabYourPitchforks** [comment]
I'm closing this in favor of https://github.com/dotnet/runtime/issues/25423, which proposes more concrete APIs.

@tannergooding - check out the original post at the top of this issue for more context. I assume your proposal argues for similar behavior.

**Comment by GrabYourPitchforks** [comment]
Bah, you're no fun. :)
